### PR TITLE
WINC-1037: Allows selecting Windows Server version for testing

### PR DESCRIPTION
**Pull request by aravindhp**
- Make Windows version configurable in the e2e tests by using an environment variable, `WINDOWS_SERVER_VERSION`.
- Make Windows version configurable in machineset.sh using a command line option

### CODE REVIEW

**Comment by aravindhp** [comment]
Tested locally with different WINDOWS_SERVER_VERSION  values, `2019` and `WRONG`.

/test aws-e2e-operator
/test azure-e2e-operator

**Comment by aravindhp** [comment]
Previous tests passed and that was with an early exit in the run-ci-e2e-test.sh script to get early feedback if the `WINDOWS_SERVER_VERSION` was getting propagated in CI. Lets now try actual tests.

/test aws-e2e-operator
/test azure-e2e-operator

**Comment by aravindhp** [comment]
/retitle WINC-1037: WIP: Allows selecting Windows Server version for testing

**Comment by aravindhp** [comment]
/retitle WINC-1037: Allows selecting Windows Server version for testing

File: test/e2e/providers/aws/aws.go
Comment (saifshaikh48) [inline test/e2e/providers/aws/aws.go:16]: nit: wrong import group

File: test/e2e/providers/cloudprovider.go
Comment (saifshaikh48) [inline test/e2e/providers/cloudprovider.go:56]: I thought we supported 2019 on platform-none, at least [according to our docs](https://docs.openshift.com/container-platform/4.12/windows_containers/understanding-windows-container-workloads.html#wmco-prerequisites-supported-6.0.0_understanding-windows-container-workloads). Or is this done because of how the golden image template is hardcoded in the release repo steps?

File: hack/machineset.sh
Comment (saifshaikh48) [inline hack/machineset.sh:317]: Please also update the usage info comment at the top of the script

File: test/e2e/windows/version.go
Comment (sebsoto) [inline test/e2e/windows/version.go:16]: I don't personally agree with introducing this package/type here.
In this case it doesn't solve any problem that a map doesn't already. Looking at the [package],(https://pkg.go.dev/k8s.io/apimachinery/pkg/util/sets#Set) I understand using it if you need to do anything related to set theory, but in my opinion its better to use the built in solution here,

File: test/e2e/providers/cloudprovider.go
Comment (sebsoto) [inline test/e2e/providers/cloudprovider.go:32]: I'm not sure if it makes the most sense to have `windowsServerVersion` be defined at CloudProvider struct creation.

What about a case where we want to spin up both a 2019 and 2022 VM? This would prevent that.
Adding the version as a parameter to `GenerateMachineSet()` gives the interface more flexibility.

**Review by sebsoto — CHANGES_REQUESTED** [review]

File: test/e2e/windows/version.go
Comment (aravindhp) [reply test/e2e/windows/version.go:16]: Why would I use a map? What do you mean by the built in solution? Do you mean a regular list?

Not sure why you think the set package should be used only if we are doing something related to set theory. This is used heavily upstream in similar scenarios. It just made the code look more elegant and we were already vendoring the library. But I can remove it if you insist.

File: test/e2e/network_test.go
Comment (alinaryan) [inline test/e2e/network_test.go:472]: Why is this a separate case if 2022 is the default?

**Review by alinaryan — COMMENTED** [review]
Thanks for working on this

File: test/e2e/network_test.go
Comment (aravindhp) [reply test/e2e/network_test.go:472]: This is for clarity.

File: test/e2e/providers/cloudprovider.go
Comment (aravindhp) [reply test/e2e/providers/cloudprovider.go]: This change is no longer needed.

File: hack/machineset.sh
Comment (aravindhp) [reply hack/machineset.sh:318]: Done.

File: test/e2e/windows/version.go
Comment (aravindhp) [reply test/e2e/windows/version.go]: While I don't agree with your view that using the set package is only for set theory, I have gone ahead and switch to using a list. If you want a map, I am going to insist on using a set 😃

File: test/e2e/providers/cloudprovider.go
Comment (aravindhp) [reply test/e2e/providers/cloudprovider.go]: I thought about it more and agree that `windowsServerVersion` should not be the `CloudProvider` struct. However introducing a mixed Windows Server version cluster in CI is going to way more complex and even laying the ground work beyond updating `GenerateMachineSet()` is beyond the scope of this PR.

**Comment by aravindhp** [comment]
/test e2e-aws-operator

[Strange failure](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/_windows-machine-config-operator/1586/pull-ci-openshift-windows-machine-config-operator-master-aws-e2e-operator/1651359039654727680)

```
hack/run-ci-e2e-test.sh -t basic
/cli/oc
Client Version: v4.2.0-alpha.0-1799-g42126f8
Kustomize Version: v4.5.7
Unable to connect to the server: dial tcp: lookup api.ci-op-ld3xcb2y-9c393.origin-ci-int-aws.dev.rhcloud.com on 172.30.0.10:53: no such host
make: *** [Makefile:198: run-ci-e2e-test] Error 1

```

**Review by saifshaikh48 — COMMENTED** [review]
/lgtm

**Comment by aravindhp** [comment]
/hold

Until #1398 merges as this PR will cause a conflict there

File: test/e2e/windows/version.go
Comment (sebsoto) [reply test/e2e/windows/version.go]: I was referring to using this pattern that we use elsewhere through the codebase
```
versions := map[string]struct{}{"2019":struct{}}
if _, exists := versions["2019"]; exists {
   ....
}
```
I spent some time looking into this more and I guess its a less common go pattern than I thought :sweat_smile:

Comment (sebsoto) [reply test/e2e/windows/version.go]: Example in our code: https://github.com/openshift/windows-machine-config-operator/blob/master/controllers/controllers.go#L254

File: test/e2e/main_test.go
Comment (mansikulkarni96) [inline test/e2e/main_test.go:86]: ofr-> for

**Review by mansikulkarni96 — CHANGES_REQUESTED** [review]

File: test/e2e/main_test.go
Comment (aravindhp) [reply test/e2e/main_test.go]: Fixed.

**Review by saifshaikh48 — COMMENTED** [review]
/lgtm

**Comment by aravindhp** [comment]
/test azure-e2e-ccm-install

Seeing a few [occurrences](https://search.ci.openshift.org/?search=multiple+manifests+for+group+config.openshift.io+kind+FeatureGate&maxAge=48h&context=1&type=build-log&name=&excludeName=&maxMatches=5&maxBytes=20971520&groupBy=job) of:
```
level=fatal msg=failed to fetch Metadata: failed to fetch dependency of "Metadata": failed to fetch dependency of "Bootstrap Ignition Config": failed to generate asset "CVO Ignore": multiple manifests for group config.openshift.io kind FeatureGate namespace  name cluster: manifests/feature_gate.yaml, openshift/99_feature-gate.yaml
Installer exit with code 1
```

**Comment by aravindhp** [comment]
/test aws-e2e-upgrade

https://github.com/openshift/release/pull/38939 has merged

**Comment by aravindhp** [comment]
/retest-required

[Failure](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_windows-machine-config-operator/1586/pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-operator/1653884192926208000) importing ovf. Known flake that's being investigated.

**Comment by aravindhp** [comment]
/test gcp-e2e-operator

```
 2023/05/04 18:23:55 failed to change ci-op-8gkqjtwy-8411c-wpzwk-e2e-hl2t4 default shell to Powershell
    create_test.go:198: 
        	Error Trace:	/go/src/github.com/openshift/windows-machine-config-operator/test/e2e/create_test.go:198
        	Error:      	Received unexpected error:
        	            	unable to change default shell of machine ci-op-8gkqjtwy-8411c-wpzwk-e2e-hl2t4: timed out waiting for the condition
        	Test:       	TestWMCO/create/Creation/ConfigMap_controller
        	Messages:   	error provisioning BYOH ConfigMap with MachineSets 
```
Network [flake](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_windows-machine-config-operator/1586/pull-ci-openshift-windows-machine-config-operator-master-gcp-e2e-operator/1654158394187059200/artifacts/gcp-e2e-operator/windows-e2e-operator-test/artifacts/pods/job-name=change-default-shell-job-zb58v/job-name=change-default-shell-job-zb58v.log):
```
change-default-shell-job-zb58v-qd7st: ssh: connect to host 10.0.128.6 port 22: Connection timed out
ssh: connect to host 10.0.128.6 port 22: Connection timed out
```

**Comment by aravindhp** [comment]
/hold

For https://github.com/openshift/release/pull/39019 to merge

**Comment by aravindhp** [comment]
/test azure-e2e-ccm-install

[Took took long](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_windows-machine-config-operator/1586/pull-ci-openshift-windows-machine-config-operator-master-azure-e2e-ccm-install/1654524843879567360) for nodes to be configured

**Comment by aravindhp** [comment]
`azure-e2e-ccm-install` is [failing](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_windows-machine-config-operator/1586/pull-ci-openshift-windows-machine-config-operator-master-azure-e2e-ccm-install/1654524843879567360) because the the installation of the container feature is [failing](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_windows-machine-config-operator/1586/pull-ci-openshift-windows-machine-config-operator-master-azure-e2e-ccm-install/1654524843879567360/artifacts/azure-e2e-ccm-install/gather-extra/artifacts/pods/openshift-windows-machine-config-operator_windows-machine-config-operator-6d6586dc84-b9p9b_manager.log):
```
2023-05-05T18:10:59Z	DEBUG	wc 10.0.128.7	run	{"cmd": "powershell.exe -NonInteractive -ExecutionPolicy Bypass \"Get-WindowsOptionalFeature -FeatureName Containers -Online\"", "out": "\r\n\r\nFeatureName      : Containers\r\nDisplayName      : Containers\r\nDescription      : Provides services and tools to create and manage Windows Server Containers and their resources.\r\nRestartRequired  : Possible\r\nState            : Disabled\r\nCustomProperties : \r\n                   ServerComponent\\Description : Provides services and tools to create and manage Windows Server \r\n                   Containers and their resources.\r\n                   ServerComponent\\DisplayName : Containers\r\n                   ServerComponent\\Id : 1018\r\n                   ServerComponent\\Type : Feature\r\n                   ServerComponent\\UniqueName : Containers\r\n                   ServerComponent\\Deploys\\Update\\Name : Containers\r\n                   ServerComponent\\Version\\Major : 10\r\n                   ServerComponent\\Version\\Minor : 0\r\n                   ServerComponent\\MutualExclusion\\MutualExclusionGroup\\Name : NetworkVirtualization\r\n                   \r\n\r\n\r\n\r\n"}
2023-05-05T18:11:02Z	ERROR	wc 10.0.128.7	error running	{"cmd": "powershell.exe -NonInteractive -ExecutionPolicy Bypass \"Install-WindowsFeature -Name Containers\"", "out": "Install-WindowsFeature : Win32 internal error \"Access is denied\" 0x5 occurred while reading the console output buffer. \r\nContact Microsoft Customer Support Services.\r\nAt line:1 char:1\r\n+ Install-WindowsFeature -Name Containers\r\n+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\r\n    + CategoryInfo          : ReadError: (:) [Install-WindowsFeature], HostException\r\n    + FullyQualifiedErrorId : ReadConsoleOutput,Microsoft.Windows.ServerManager.Commands.AddWindowsFeatureCommand\r\n \r\n", "error": "Process exited with status 1"}
github.com/openshift/windows-machine-config-operator/pkg/windows.(*windows).Run
	/build/windows-machine-config-operator/pkg/windows/windows.go:373
github.com/openshift/windows-machine-config-operator/pkg/windows.(*windows).enableContainersWindowsFeature
	/build/windows-machine-config-operator/pkg/windows/windows.go:893
github.com/openshift/windows-machine-config-operator/pkg/windows.(*windows).ensureHostNameAndContainersFeature
	/build/windows-machine-config-operator/pkg/windows/windows.go:530
github.com/openshift/windows-machine-config-operator/pkg/windows.(*windows).Bootstrap
	/build/windows-machine-config-operator/pkg/windows/windows.go:430
github.com/openshift/windows-machine-config-operator/pkg/nodeconfig.(*nodeConfig).Configure
	/build/windows-machine-config-operator/pkg/nodeconfig/nodeconfig.go:159
github.com/openshift/windows-machine-config-operator/controllers.(*instanceReconciler).ensureInstanceIsUpToDate
	/build/windows-machine-config-operator/controllers/controllers.go:84
github.com/openshift/windows-machine-config-operator/controllers.(*WindowsMachineReconciler).configureMachine
	/build/windows-machine-config-operator/controllers/windowsmachine_controller.go:433
github.com/openshift/windows-machine-config-operator/controllers.(*WindowsMachineReconciler).Reconcile
	/build/windows-machine-config-operator/controllers/windowsmachine_controller.go:354
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
	/build/windows-machine-config-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:122
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/build/windows-machine-config-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:323
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/build/windows-machine-config-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:274
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/build/windows-machine-config-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:235
```
Needs investigation if this has suddenly become an issue on Windows Server 2019 Azure VMs.

File: test/e2e/providers/azure/azure.go
Comment (saifshaikh48) [inline test/e2e/providers/azure/azure.go:135]: The `with-containers` images are getting removed any time now. Azure jobs testing with 2019 would be blocked. Should we explore alternatives before introducing this i.e. moving 2019 testing away from azure?

Comment (aravindhp) [reply test/e2e/providers/azure/azure.go:135]: I am aware of that but there is no easy fix that I can add, which is why I created [OCPBUGS-13244](https://issues.redhat.com/browse/OCPBUGS-13244), otherwise this is going to hold up the AWS 2022 support. Worst case if tests start failing, we can flip back to 2022 for Azure in the release repo.

**Comment by aravindhp** [comment]
/retest-required

[nutanix-e2e-operator](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_windows-machine-config-operator/1586/pull-ci-openshift-windows-machine-config-operator-master-nutanix-e2e-operator/1655975674948620288) is failing during bootstrapping. Seeing errors in the kubelet.log when bringing up the `keepalived-monitor` but not sure if this is the root cause:

**Comment by aravindhp** [comment]
/retest-required

Same failure again

**Comment by aravindhp** [comment]
/test nutanix-e2e-operator

The test continues to fail during cluster install. If this keeps up it warrants a bug report. Cc: @yanhua121

**Comment by aravindhp** [comment]
/retest-required

The nutanix-e2e-operator job was failing due to stale VMs from previous runs

**Comment by aravindhp** [comment]
/retest-required

🤞🏽

**Pull request merged**
